### PR TITLE
py-graveyard/p5-graveyard: add support for epoch

### PR DESCRIPTION
**Pull request by reneeotten**
#### Description
As noted by @ryandesign in [this ticket](https://trac.macports.org/ticket/56858), replacements don't work correctly if the ```epoch``` value is higher than ```0```. This PR fixes that and adds support for ```epoch``` to py-graveyard and p5-graveyard. In addition, it adds the epoch value for py26-webhelpers (see [this commit](https://github.com/macports/macports-ports/commit/aa1d608282a2a58b3b9d3c4396a361832879448d#comments)).

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.13.6 17G65
Xcode 9.4.1 9F2000

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by reneeotten** [comment]
@pmetzger can you please request a review from @ryandesign and @larryv ?

**Comment by pmetzger** [comment]
@reneeotten Requested.

**Comment by mojca** [comment]
I did not test, but it looks OK.

**Comment by pmetzger** [comment]
@reneeotten Let me know when you think we've waited long enough to merge this.

Anyone else, please have a look and comment!

**Comment by reneeotten** [comment]
I have tested it for ```py-graveyard``` and am, therefore, fairly confident that this is a correct solution.  I never used any of the perl stuff, but I assume that it will work there as well since it's implemented in the same way (but perhaps @dbevans could take a look?).

**Comment by ryandesign** [comment]
> In addition, it adds the epoch value for py26-webhelpers

This is probably not the only one that needs an epoch added. From the python directory of the ports tree, I ran:

    grep -El '^[[:space:]]*epoch' */Portfile | cut -d / -f 1 | xargs echo | tr ' ' '|' | xargs -J % grep -E % py-graveyard/Portfile | awk '{print "* " $1}'

The following python ports that have entries in py-graveyard currently use an epoch; if they used an epoch at the time that they were added to py-graveyard, then they need that epoch added to py-graveyard:

* py-asn1
* py-gdbm
* py-gmpy2
* py-gobject
* py-gobject3
* py-graph-tool
* py-pybtex
* py-soaplib
* py-spyder
* py-spyder-devel
* py-tkinter
* py-vobject
* py-voeventlib
* py-webhelpers

Similarly, from within the perl directory, I ran:

    grep -El '^[[:space:]]*epoch' */Portfile | cut -d / -f 1 | xargs echo | tr ' ' '|' | xargs -J % grep -E % p5-graveyard/Portfile | awk '{print "* " $1}'

And the perl ports it shows which have entries in p5-graveyard and which are currently using an epoch are:

* p5-array-compare
* p5-catalyst-plugin-authentication
* p5-catalyst-plugin-authentication-store-dbic
* p5-chemistry-elements
* p5-datetime
* p5-datetime-calendar-christian
* p5-datetime-calendar-julian
* p5-datetime-format-builder
* p5-datetime-format-dateparse
* p5-datetime-format-iso8601
* p5-datetime-format-mail
* p5-datetime-format-mysql
* p5-datetime-format-pg
* p5-datetime-format-sqlite
* p5-datetime-format-strptime
* p5-datetime-format-w3cdtf
* p5-datetime-locale
* p5-datetime-timezone
* p5-devel-stacktrace
* p5-devel-stacktrace-ashtml
* p5-extutils-cbuilder
* p5-html-tree
* p5-html-treebuilder-xpath
* p5-ifeffit
* p5-io-aio
* p5-json-xs
* p5-module-build
* p5-module-build-ffi
* p5-module-build-tiny
* p5-module-build-xsutil
* p5-mojolicious
* p5-mojolicious-plugin-chi
* p5-mojolicious-plugin-mason1renderer
* p5-mojolicious-plugin-mason2renderer
* p5-mojolicious-plugin-setusergroup
* p5-parse-yapp
* p5-perl-critic
* p5-perlmagick
* p5-spreadsheet-parseexcel
* p5-sub-uplevel
* p5-test-differences
* p5-time-epoch
* p5-version
* p5-xml-autowriter
* p5-xml-feed

**Comment by ryandesign** [comment]
> What I'd probably do is: If the specification has one number, it's unambiguous: that one number is the version. If it has three numbers, it's also unambiguous: it's epoch_version_revision. If it has two numbers, it's either epoch_version or version_revision, and the way I would differentiate between them is that whichever one is not an integer is the version. In the unlikely event that both are integers, then treat it as version_revision.

Alternately, it could be made unambiguous by using a different separator for the epoch, for example epoch/version_revision.

**Comment by ryandesign** [comment]
> @ryandesign thank you for the comments Ryan. On the one hand you approve the PR, but then also suggest to make changes - so, what would you prefer me to do here?

Pull request etiquette still isn't 100% clear to me. On the one hand, your changes look like they should work the way they are, so on that basis I don't want to reject the PR. But I would prefer to see a change to have the epoch listed first, not last.

> If you want to have it listed in order of precedence, then I would personally go for the epoch/version_revision scheme, just because for the other option my Tcl knowledge is probably insufficient...

Using a different separator for the epoch is probably best, to avoid ambiguity and to simplify the code.

We could use the separator `@` instead of `/`. We already use the notation `@version_revision` elsewhere, like in the output of `port info` and `port installed`. So then here it would become `epoch@version_revision`.

> If you want me to change it accordingly, just let me know and I will give it a try and update this PR. After the exact syntax is settled, I can take a look at the python ports you listed and make sure that the epoch is added correctly for all of them. I am probably not motivated enough to do so for perl ports though.

I can do the perl ports.

**Comment by ryandesign** [comment]
There are fewer affected ports than I originally reported. I've fixed my scripts and their output above.

**Comment by mojca** [comment]
There is one more thing that I would like to see fixed for perl. On top of `p5-graveyard` there is a command used to generate the list of ports. I always run just that command and never do any manual fixes. That command needs to be changed to include the epoch in the correct format.

**Comment by ryandesign** [comment]
> I can do the perl ports.

It's easy enough to automate, so I've done the python ports too. The only weird ones that needed special care were py-gdbm and py-tkinter, which only use epoch on py33 and later subports, not on the versions in the graveyard.

> On top of p5-graveyard there is a command used to generate the list of ports. I always run just that command and never do any manual fixes. That command needs to be changed to include the epoch in the correct format.

I forgot about this! Thanks for pointing it out. I'll look into updating that too.

**Comment by ryandesign** [comment]
We'll have to squash some of these commits later but I didn't want to do that while multiple people are working on it. But I'm finished working on it for now. Renee, can you handle the rest?

File: perl/p5-graveyard/Portfile [hunk 1]
```diff
@@ -49,21 +53,27 @@ proc unknown args {
...
 #               printf "\t"
-#           printf "%s_%d%s", $1, ($2 + 1), ORS
+#           printf "%s%s_%d%s", epoch ? epoch "@" : "", version, revision + 1, ORS
 #       }
 #   '
 
 #                                                               REPLACED BRANCHES
 #                                                               (optional, defaults to
-# SUPERPORT                                     VERSION_REV     "5.24")
+# SUPERPORT                                     VER_REV_EPO     "5.24")
```
Comment (mojca) [inline perl/p5-graveyard/Portfile]: Shouldn't this be EPO@VER_REV?

File: python/py-graveyard/Portfile [hunk 2]
```diff
@@ -20,32 +20,36 @@ array set replacements {
...
     }
 }
 
 # Please keep the list of ports sorted in alphabetical order. When adding a port
-# to the list, the version_revision listed must be greater than it was before,
+# to the list, the version_revision_epoch listed must be greater than it was before,
 # otherwise the port is not considered outdated and replacement does not occur.
 
-# SUPERPORT             VERSION_REV REPLACED BRANCHES
+# SUPERPORT             VER_REV_EPO REPLACED BRANCHES
```
Comment (mojca) [inline python/py-graveyard/Portfile]: Same comment as above.

File: perl/p5-graveyard/Portfile [hunk 3]
```diff
@@ -18,16 +18,20 @@ proc unknown args {
         lappend args {*}$defaultBranches
     }
     set replaced_branches [lassign $args superport version_string]
-    lassign [split $version_string _] ver rev
+    lassign [split $version_string _] ver rev epo
```
Comment (mojca) [inline perl/p5-graveyard/Portfile]: How does this work with the `@` sign?

Comment (reneeotten) [reply perl/p5-graveyard/Portfile]: @mojca regarding your comments above: @ryandesign already updated the lists of ports with the final naming scheme, but the actual code to parse this is still what I initially suggested. I will change this accordingly and update this PR - for now it indeed does not work ;)

**Comment by reneeotten** [comment]
@ryandesign @mojca how about this? The last two commits implement the agreed upon ```epoch@version_revision``` scheme and work for me, but perhaps there is a better way of doing this in Tcl... If you approve, feel free to squash some commits (or tell me to do it) and then this PR should be ready to merge.

**Comment by reneeotten** [comment]
okay, I already squashed the commits so that there is one per Portfile.

**Comment by pmetzger** [comment]
My only comment: there should be a comment in the perl graveyard that, like the one in the python graveyard, explains the content and the @ convention.

**Comment by reneeotten** [comment]
@pmetzger done, added a similar comment to p5-graveyard

**Comment by pmetzger** [comment]
LGTM. Shall we merge?

**Comment by reneeotten** [comment]
fine with me, but I was hoping for a final approval by @ryandesign

**Comment by pmetzger** [comment]
We'll give @ryandesign a few more hours, but he might be overwhelmed.

**Comment by pmetzger** [comment]
Merged. Thank you, @reneeotten. You might want to post to the developers list noting the new format.

**Pull request merged**